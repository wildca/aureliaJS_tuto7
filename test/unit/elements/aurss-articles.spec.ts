import { bootstrap } from 'aurelia-bootstrapper';
import {PLATFORM} from 'aurelia-pal';
import { StageComponent } from 'aurelia-testing';
import {createArticle, createReadArticle} from '../../__fixtures__/articles-fixtures';
import {prepareI18nComponent, selectElementWithText} from '../../helpers';

describe('aurss-article', () => {
    let component;
    let viewModel;

    beforeEach(async(done) => {
        viewModel = {
            article: createArticle(),
            markArticleAsRead: jest.fn(),
            markArticleAsUnread: jest.fn(),
            openArticle: jest.fn(),
        };

        component = StageComponent.withResources(PLATFORM.moduleName('../../src/resources/elements/aurss-article'))
            .inView(`
            <aurss-article
                value.bind="article"
                mark-article-as-read.bind="markArticleAsRead"
                mark-article-as-unread.bind="markArticleAsUnread"
                open-article.bind="openArticle"
            ></aurss-article>
            `).boundTo(viewModel);

        prepareI18nComponent(component);

        await component.create(bootstrap);

        done();
    });

    afterEach(() => {
        component.dispose();
    });

    it('should render and match snapshot', () => {
        expect(document.body.outerHTML).toMatchSnapshot();
    });

    it('should call mark as read when mark as read button is clicked', () => {
        const markAsReadButton = selectElementWithText(document, 'button', 'Mark as read');

        markAsReadButton.click();

        expect(viewModel.markArticleAsRead).toHaveBeenCalledWith(viewModel.article);
    });

    it('should call mark as unread when mark as unread button is clicked', async() => {
        // We must mark as read first so the button is in the view model
        // and wait for the change to be applied.
        viewModel.article = createReadArticle();
        await component.waitForElement('.read', { present: true, timeout: 500 });
        const markAsUnreadButton = selectElementWithText(document, 'button', 'Mark as unread');

        markAsUnreadButton.click();

        expect(viewModel.markArticleAsUnread).toHaveBeenCalledWith(viewModel.article);
    });

    it('should open when the open button is clicked', () => {
        const openButton = selectElementWithText(document, 'button', 'Open');
        const focus = jest.fn();
        window.open = jest.fn().mockReturnValue({ focus });

        openButton.click();

        expect(viewModel.openArticle).toHaveBeenCalledWith(viewModel.article);
    });
});
