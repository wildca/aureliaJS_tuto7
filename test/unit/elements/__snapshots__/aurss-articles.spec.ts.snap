// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`aurss-article should render and match snapshot 1`] = `
"<body><div>
            <aurss-article value.bind=\\"article\\" mark-article-as-read.bind=\\"markArticleAsRead\\" mark-article-as-unread.bind=\\"markArticleAsUnread\\" open-article.bind=\\"openArticle\\" class=\\"au-target\\" au-target-id=\\"18\\">
    <div class=\\"au-target tile is-parent unread\\" au-target-id=\\"1\\">
        <div class=\\"tile is-child\\">
            <div class=\\"card\\">
                <header class=\\"card-content\\">
                    <div class=\\"is-pulled-right\\">
                        <!--anchor-->
                        <a click.delegate=\\"markAsFavorite()\\" class=\\"au-target\\" au-target-id=\\"4\\">
                            <i class=\\"far fa-star is-size-5\\"></i>
                        </a><!--anchor-->
                    </div>
                    <p class=\\"title\\">
                        <!-- We use mouseup instead of click to correctly detect wheel clicks. -->
                        <a href.bind=\\"value.link\\" target=\\"_blank\\" mouseup.delegate=\\"markAsRead()\\" class=\\"au-target\\" au-target-id=\\"6\\" href=\\"https://example.com/article/0\\">Test article</a>
                    </p>
                    <p class=\\"subtitle is-4\\">Me 6/10/2019, 13:00 â€“ Test feed</p>
                </header>
                <footer class=\\"card-footer\\">
                    <div class=\\"card-footer-item is-hidden-mobile\\">
                        <!--anchor-->
                        <button class=\\"button is-light is-fullwidth au-target\\" type=\\"button\\" click.delegate=\\"markAsFavorite()\\" t=\\"article-mark-as-favorite-btn\\" au-target-id=\\"11\\"></button><!--anchor-->
                    </div>
                    <div class=\\"card-footer-item is-hidden-mobile\\">
                        <!--anchor-->
                            <button class=\\"button is-light is-fullwidth au-target\\" type=\\"button\\" click.delegate=\\"markAsRead()\\" t=\\"article-mark-as-read-btn\\" au-target-id=\\"15\\">
                            Mark as read
                        </button><!--anchor-->
                    </div>
                    <div class=\\"card-footer-item\\">
                        <button class=\\"button is-primary is-fullwidth au-target\\" type=\\"button\\" click.delegate=\\"open()\\" t=\\"article-open-btn\\" au-target-id=\\"17\\">
                            Open
                        </button>
                    </div>
                </footer>
            </div>
        </div>
    </div>
</aurss-article>
            </div></body>"
`;
